addon/utils/decorator-macros.js

```diff
@@ -44,3 +44,25 @@ export function decoratedPropertyWithOptionalCallback(fn) {
     return fn(...params, value);
   });
 }
+
+export function decoratedPropertyWithEitherCallbackOrProperty(fn) {
+  return decoratorWithParams(function(target, key, desc, params) {
+    assert(`Cannot use '${fn.name}' on field '${key}' without parameters`, params.length !== 0);
+
+    const lastParam = params[params.length - 1]
+
+    if (typeof lastParam === 'function') {
+      return fn(...params);
+    }
+
+    if (params.length > 1 && typeof lastParam === 'string') {
```

Comment:
Seems like we are calling `typeof lastParam` twice, would you mind extracting to a separate variable to track (e.g. `let lastParamType = typeof lastParam`)